File: conformance/tests/token-ft/token-ft.runar.zig
const runar = @import("runar");

pub const FungibleToken = struct {
    pub const Contract = runar.StatefulSmartContract;

    owner: runar.PubKey = "000000000000000000000000000000000000000000000000000000000000000000",
    balance: i64 = 0,
    mergeBalance: i64 = 0,
    tokenId: runar.ByteString,

    pub fn init(owner: runar.PubKey, balance: i64, mergeBalance: i64, tokenId: runar.ByteString) FungibleToken {
        return .{
            .owner = owner,
            .balance = balance,
            .mergeBalance = mergeBalance,
            .tokenId = tokenId,
        };
    }

    pub fn transfer(
        self: *FungibleToken,
        ctx: runar.StatefulContext,
        sig: runar.Sig,
        to: runar.PubKey,
        amount: i64,
        outputSatoshis: i64,
    ) void {
        runar.assert(runar.checkSig(sig, self.owner));
        runar.assert(outputSatoshis >= 1);
        const totalBalance = self.balance + self.mergeBalance;
        runar.assert(amount > 0);
        runar.assert(amount <= totalBalance);

        ctx.addOutput(outputSatoshis, .{ to, amount, 0 });
        if (amount < totalBalance) {
            ctx.addOutput(outputSatoshis, .{ self.owner, totalBalance - amount, 0 });
        }
    }

    pub fn send(self: *FungibleToken, ctx: runar.StatefulContext, sig: runar.Sig, to: runar.PubKey, outputSatoshis: i64) void {
        runar.assert(runar.checkSig(sig, self.owner));
        runar.assert(outputSatoshis >= 1);
        ctx.addOutput(outputSatoshis, .{ to, self.balance + self.mergeBalance, 0 });
    }

    pub fn merge(
        self: *FungibleToken,
        ctx: runar.StatefulContext,
        sig: runar.Sig,
        otherBalance: i64,
        allPrevouts: runar.ByteString,
        outputSatoshis: i64,
    ) void {
        runar.assert(runar.checkSig(sig, self.owner));
        runar.assert(outputSatoshis >= 1);
        runar.assert(otherBalance >= 0);
        runar.assert(runar.bytesEq(runar.hash256(allPrevouts), runar.extractHashPrevouts(ctx.txPreimage)));

        const myOutpoint = runar.extractOutpoint(ctx.txPreimage);
        const firstOutpoint = runar.substr(allPrevouts, 0, 36);
        const myBalance = self.balance + self.mergeBalance;

        if (runar.bytesEq(myOutpoint, firstOutpoint)) {
            ctx.addOutput(outputSatoshis, .{ self.owner, myBalance, otherBalance });
        } else {
            ctx.addOutput(outputSatoshis, .{ self.owner, otherBalance, myBalance });
        }
    }
};

File: conformance/tests/token-ft/token-ft.runar.go
//go:build ignore

package contract

import "runar"

type FungibleToken struct {
	runar.StatefulSmartContract
	Owner        runar.PubKey
	Balance      runar.Bigint
	MergeBalance runar.Bigint
	TokenId      runar.ByteString `runar:"readonly"`
}

func (c *FungibleToken) Transfer(sig runar.Sig, to runar.PubKey, amount runar.Bigint, outputSatoshis runar.Bigint) {
	runar.Assert(runar.CheckSig(sig, c.Owner))
	runar.Assert(outputSatoshis >= 1)
	totalBalance := c.Balance + c.MergeBalance
	runar.Assert(amount > 0)
	runar.Assert(amount <= totalBalance)
	c.AddOutput(outputSatoshis, to, amount, 0)
	if amount < totalBalance {
		c.AddOutput(outputSatoshis, c.Owner, totalBalance-amount, 0)
	}
}

func (c *FungibleToken) Send(sig runar.Sig, to runar.PubKey, outputSatoshis runar.Bigint) {
	runar.Assert(runar.CheckSig(sig, c.Owner))
	runar.Assert(outputSatoshis >= 1)
	c.AddOutput(outputSatoshis, to, c.Balance+c.MergeBalance, 0)
}

func (c *FungibleToken) Merge(sig runar.Sig, otherBalance runar.Bigint, allPrevouts runar.ByteString, outputSatoshis runar.Bigint) {
	runar.Assert(runar.CheckSig(sig, c.Owner))
	runar.Assert(outputSatoshis >= 1)
	runar.Assert(otherBalance >= 0)
	runar.Assert(runar.Hash256(allPrevouts) == runar.ExtractHashPrevouts(c.TxPreimage))
	myOutpoint := runar.ExtractOutpoint(c.TxPreimage)
	firstOutpoint := runar.Substr(allPrevouts, 0, 36)
	myBalance := c.Balance + c.MergeBalance
	if myOutpoint == firstOutpoint {
		c.AddOutput(outputSatoshis, c.Owner, myBalance, otherBalance)
	} else {
		c.AddOutput(outputSatoshis, c.Owner, otherBalance, myBalance)
	}
}
